SU2_CFD/src/numerics_direct_scalar.cpp

```diff
@@ -0,0 +1,390 @@
+  AD::SetPreaccIn(Normal, nDim);
+  AD::SetPreaccIn(ScalarVar_Grad_i, nVar, nDim);
+  AD::SetPreaccIn(ScalarVar_Grad_j, nVar, nDim);
+  AD::SetPreaccIn(Diffusion_Coeff_i, nVar);
+  AD::SetPreaccIn(Diffusion_Coeff_j, nVar);
+  if (correct_gradient) {
+    AD::SetPreaccIn(ScalarVar_i, nVar); AD::SetPreaccIn(ScalarVar_j, nVar);
+  }
+  ExtraADPreaccIn();
+  
+  if (incompressible) {
+    AD::SetPreaccIn(V_i, nDim+6); AD::SetPreaccIn(V_j, nDim+6);
+    
+    Density_i           = V_i[nDim+2];            Density_j = V_j[nDim+2];
+    Laminar_Viscosity_i = V_i[nDim+4];  Laminar_Viscosity_j = V_j[nDim+4];
+    Eddy_Viscosity_i    = V_i[nDim+5];     Eddy_Viscosity_j = V_j[nDim+5];
+  }
+  else {
+    AD::SetPreaccIn(V_i, nDim+7); AD::SetPreaccIn(V_j, nDim+7);
+    
+    Density_i           = V_i[nDim+2];            Density_j = V_j[nDim+2];
+    Laminar_Viscosity_i = V_i[nDim+5];  Laminar_Viscosity_j = V_j[nDim+5];
+    Eddy_Viscosity_i    = V_i[nDim+6];     Eddy_Viscosity_j = V_j[nDim+6];
+  }
+  
+  /*--- Compute vector going from iPoint to jPoint ---*/
+  
+  dist_ij_2 = 0; proj_vector_ij = 0;
+  for (iDim = 0; iDim < nDim; iDim++) {
+    Edge_Vector[iDim] = Coord_j[iDim]-Coord_i[iDim];
+    dist_ij_2 += Edge_Vector[iDim]*Edge_Vector[iDim];
+    proj_vector_ij += Edge_Vector[iDim]*Normal[iDim];
+  }
+  if (dist_ij_2 == 0.0) proj_vector_ij = 0.0;
+  else proj_vector_ij = proj_vector_ij/dist_ij_2;
+  
+  /*--- Mean gradient approximation ---*/
+  for (iVar = 0; iVar < nVar; iVar++) {
+    Proj_Mean_GradScalarVar_Normal[iVar] = 0.0;
+    Proj_Mean_GradScalarVar_Edge[iVar] = 0.0;
+    for (iDim = 0; iDim < nDim; iDim++) {
+      Mean_GradScalarVar[iVar][iDim] = 0.5*(ScalarVar_Grad_i[iVar][iDim] +
+                                            ScalarVar_Grad_j[iVar][iDim]);
+      Proj_Mean_GradScalarVar_Normal[iVar] += Mean_GradScalarVar[iVar][iDim] *
+      Normal[iDim];
+      if (correct_gradient)
+        Proj_Mean_GradScalarVar_Edge[iVar] += Mean_GradScalarVar[iVar][iDim]*Edge_Vector[iDim];
+    }
+    Proj_Mean_GradScalarVar[iVar] = Proj_Mean_GradScalarVar_Normal[iVar];
+    if (correct_gradient) {
+      Proj_Mean_GradScalarVar[iVar] -= Proj_Mean_GradScalarVar_Edge[iVar]*proj_vector_ij -
+      (ScalarVar_j[iVar]-ScalarVar_i[iVar])*proj_vector_ij;
+    }
+  }
+  
+  FinishResidualCalc(val_residual, Jacobian_i, Jacobian_j, config);
+  
+  AD::SetPreaccOut(val_residual, nVar);
+  AD::EndPreacc();
+  
+}
+
+CAvgGradScalar_General::CAvgGradScalar_General(unsigned short val_nDim,
+                                               unsigned short val_nVar, bool correct_grad,
+                                               CConfig *config)
+: CAvgGradScalar(val_nDim, val_nVar, correct_grad, config) {
+  
+  Mean_Diffusivity = new su2double[nVar];
+  
+}
+
+CAvgGradScalar_General::~CAvgGradScalar_General(void) {
+  
+  if (Mean_Diffusivity != NULL) delete [] Mean_Diffusivity;
+  
+}
+
+void CAvgGradScalar_General::ExtraADPreaccIn() { }
```

Comment:
I can chime in here.  In the original implementation, the "scalars" were the turbulent variables.  I was the one who applied the [template design pattern](https://sourcemaking.com/design_patterns/template_method) to the `AvgGrad` classes.

The motivation was this: In the template method, you find what's the same and what's different between similar functions.  You put everything that's the same into a base class (e.g. `CAvgGradScalar`), then put everything that differs into functions implemented by the derived class.  So some `SetPreaccIn` calls are common, and they're in the base class (`CAvgGradScalar`).  Those that are unique are included in the `ExtraADPreaccIn`.  That way the `CAvgGradScalar::ComputeResidual` just does this for all child classes:

```cpp
void CAvgGradScalar::ComputeResidual(su2double *val_residual,
                                     su2double **Jacobian_i,
                                     su2double **Jacobian_j,
                                     CConfig *config) {

  AD::StartPreacc();
  AD::SetPreaccIn(Coord_i, nDim); AD::SetPreaccIn(Coord_j, nDim);
  AD::SetPreaccIn(Normal, nDim);
  AD::SetPreaccIn(ScalarVar_Grad_i, nVar, nDim);
  AD::SetPreaccIn(ScalarVar_Grad_j, nVar, nDim);
  AD::SetPreaccIn(Diffusion_Coeff_i, nVar);
  AD::SetPreaccIn(Diffusion_Coeff_j, nVar);
  if (correct_gradient) {
    AD::SetPreaccIn(ScalarVar_i, nVar); AD::SetPreaccIn(ScalarVar_j, nVar);
  }
  ExtraADPreaccIn();
```

Now, if you're OK with code duplication, then *all* the `SetPreaccIn` calls could be made in derived classes.  But it's impossible to put them all in the `ComputeResidual` class because not all derived classes use the same variables.

Reply:
I see (I've overseen the `_General` extension to spot the overload), thanks for the explanation! I didn't notice the new templated structure..
Problems due to erroneous preaccumulation happened quite often to me (and they are very hard to track down), so that's why I'm so worried about those lines :D - though it's not reason enough for duplicating the code.

Reply:
If you missed it, then I would recommend that @economon add some explanatory comments about `CAvgGradScalar_General` either in the header file or the cpp file.  That way future code readers/programmers don't have the same confusion.

Unless @economon would prefer that I do the comments.

Reply:
Thanks for highlighting this topic. I think @oleburghardt's experience is very common. I had the same issue: the adjoint was not working with scalars for a while before I realized I had made the same mistake of not adding the diffusivity to the preaccumulation. This is very easy to overlook.

Is there a change we can make that fits within the newer class structure of @clarkpede to make this more clear? Comments are great, but they become stale over time, and folks may not read it. It would be good to force folks to consider this somehow while writing the routines. Is there a clever way to error check this for instance?